### ISSUE TITLE
xrdp crashes after installing Let's Encrypt certificate

### ISSUE THREAD

**Issue by iblackwell** [neutrinolabs/xrdp#2434]
I have a fresh LAMP system, with SSL enabled in apache2 using the (default) snakeoil certificate. I have been heretofore able to reliably access the system via xrdp. Today I replaced the snakeoil certificate with a Let's Encrypt certificate using certbot. That was the only change I made. Immediately thereafter, any xrdp session I try to initiate crashes a minute or two after authentication. Specifically, the client window just disappears. Having been summarily booted out by xrdp, I am unable to access the server either with xrdp (not window opens and closes immediately) or at the server console (get a blank screen after entering my credentials). When I go to reboot, I'm told that someone is logged in.

Interestingly, if I boot the server, stop the xrdp service (sudo), start the service again and log out, I can then access the server by xrdp and it is stable. Not only is that a workaround, but it should say something about the nature of the issue (but that'd be above my pay-grade).

System info:
Ubuntu 22.04.1
apache2-2.4.52-1ubuntu4.2
mariadb-serverp-10.6.7-2ubuntu1.1
php-8.1+92ubuntu1
xrdp-0.9.17-2ubuntu2

There are three log files copied below:
**xrdp.log-stable** - Log file from stable session prior to installing Let's Encrypt cert. using certbot
**xrdp.log-unstable** - Log file from crashed session after installing Let's Encrypt cert
**xrdp-sesman.log-unstable** - Session log file from crashed session

Highlights from **xrdp.log-unstable.txt**:
14:03:32 - I initiated a remote xrdp session
14:03:45 - I got spanked for entering an incorrect password
14:03:50 - I was admitted into the xrdp session (having supplied the correct password)
14:05:00 - xrdp window disappears from my screen
14:05:41 - xrdp resets itself for another session

Other notes:
1. "Brian" (192.168.1.100) referred to in the logs is my client workstation that uses xfreerdp or remmina to access the server. Both clients crash in the same manner.
2. I use tcp/14403 for xrdp in the server.

Any insight into what might be the problem here would be appreciated. Thanks much in advance.

**xrdp.log-stable**
```
[20221116-12:37:29] [INFO ] starting xrdp with pid 964
[20221116-12:37:29] [INFO ] address [0.0.0.0] port [14403] mode 1
[20221116-12:37:29] [INFO ] listening to port 14403 on 0.0.0.0
[20221116-12:37:29] [INFO ] xrdp_listen_pp done
[20221116-12:38:14] [INFO ] Socket 12: AF_INET6 connection received from ::ffff:192.168.1.100 port 35958
[20221116-12:38:14] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20221116-12:38:14] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20221116-12:38:14] [ERROR] Cannot read private key file /etc/xrdp/key.pem: Permission denied
[20221116-12:38:14] [INFO ] Connected client computer name: Brian
[20221116-12:38:14] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc006 is unknown (ignored)
[20221116-12:38:14] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc00a is unknown (ignored)
[20221116-12:38:14] [INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x04], k
eyboard_subtype:[0x00], keylayout:[0x00000409]
[20221116-12:38:14] [INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[20221116-12:38:14] [INFO ] Non-TLS connection established from ::ffff:192.168.1.100 port 35958: encrypted with standard RDP security
[20221116-12:38:15] [INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[20221116-12:38:15] [INFO ] xrdp_caps_process_codecs: RemoteFX, codec id 3, properties len 49
[20221116-12:38:15] [INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[20221116-12:38:15] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
[20221116-12:38:15] [ERROR] dynamic_monitor_open_response: error
[20221116-12:38:15] [ERROR] xrdp_rdp_recv: xrdp_channel_process failed
[20221116-12:38:19] [INFO ] connecting to sesman ip 127.0.0.1 port 3350
[20221116-12:38:19] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20221116-12:38:19] [INFO ] sesman connect ok
[20221116-12:38:19] [INFO ] sending login info to session manager, please wait...
[20221116-12:38:19] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20221116-12:38:19] [INFO ] login successful for display 10
[20221116-12:38:19] [INFO ] loaded module 'libxup.so' ok, interface size 10296, version 4
[20221116-12:38:19] [INFO ] started connecting
[20221116-12:38:19] [INFO ] lib_mod_connect: connecting via UNIX socket
[20221116-12:38:20] [INFO ] lib_mod_log_peer: xrdp_pid=1691 connected to X11rdp_pid=1707 X11rdp_uid=1000 X11rdp_gid=1000 cl
ient_ip=::ffff:192.168.1.100 client_port=35958
[20221116-12:38:20] [INFO ] connected ok
```

**xrdp.log-unstable**
```
[20221116-14:03:30] [INFO ] address [0.0.0.0] port [14403] mode 1
[20221116-14:03:30] [INFO ] listening to port 14403 on 0.0.0.0
[20221116-14:03:30] [INFO ] xrdp_listen_pp done
[20221116-14:03:32] [INFO ] starting xrdp with pid 956
[20221116-14:03:32] [INFO ] address [0.0.0.0] port [14403] mode 1
[20221116-14:03:32] [INFO ] listening to port 14403 on 0.0.0.0
[20221116-14:03:32] [INFO ] xrdp_listen_pp done
[20221116-14:03:38] [INFO ] Socket 12: AF_INET6 connection received from ::ffff:192.168.1.100 port 54044
[20221116-14:03:38] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20221116-14:03:38] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20221116-14:03:38] [INFO ] Connected client computer name: Brian
[20221116-14:03:38] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc006 is unknown (ignored)
[20221116-14:03:38] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc00a is unknown (ignored)
[20221116-14:03:39] [INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x04], keyboard_subtype:[0x00], keylayout:[0x00000409]
[20221116-14:03:39] [INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[20221116-14:03:39] [INFO ] TLS connection established from ::ffff:192.168.1.100 port 54044: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384
[20221116-14:03:39] [INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[20221116-14:03:39] [INFO ] xrdp_caps_process_codecs: RemoteFX, codec id 3, properties len 49
[20221116-14:03:39] [INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[20221116-14:03:39] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
[20221116-14:03:40] [ERROR] dynamic_monitor_open_response: error
[20221116-14:03:40] [ERROR] xrdp_rdp_recv: xrdp_channel_process failed
[20221116-14:03:43] [INFO ] connecting to sesman ip 127.0.0.1 port 3350
[20221116-14:03:43] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20221116-14:03:43] [INFO ] sesman connect ok
[20221116-14:03:43] [INFO ] sending login info to session manager, please wait...
[20221116-14:03:45] [INFO ] xrdp_wm_log_msg: login failed for display 0
[20221116-14:03:45] [INFO ] login failed for display 0
[20221116-14:03:50] [INFO ] connecting to sesman ip 127.0.0.1 port 3350
[20221116-14:03:50] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20221116-14:03:50] [INFO ] sesman connect ok
[20221116-14:03:50] [INFO ] sending login info to session manager, please wait...
[20221116-14:03:50] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20221116-14:03:50] [INFO ] login successful for display 10
[20221116-14:03:50] [INFO ] loaded module 'libxup.so' ok, interface size 10296, version 4
[20221116-14:03:50] [INFO ] started connecting
[20221116-14:03:50] [INFO ] lib_mod_connect: connecting via UNIX socket
[20221116-14:03:50] [INFO ] lib_mod_log_peer: xrdp_pid=1644 connected to X11rdp_pid=1664 X11rdp_uid=1000 X11rdp_gid=1000 client_ip=::ffff:192.168.1.100 client_port=54044
[20221116-14:03:50] [INFO ] connected ok
[20221116-14:05:00] [INFO ] Received termination signal, stopping the server accept new connections thread
[20221116-14:05:41] [INFO ] address [0.0.0.0] port [14403] mode 1
[20221116-14:05:41] [INFO ] listening to port 14403 on 0.0.0.0
[20221116-14:05:41] [INFO ] xrdp_listen_pp done
[20221116-14:05:43] [INFO ] starting xrdp with pid 934
[20221116-14:05:43] [INFO ] address [0.0.0.0] port [14403] mode 1
[20221116-14:05:43] [INFO ] listening to port 14403 on 0.0.0.0
[20221116-14:05:43] [INFO ] xrdp_listen_pp done
```

**xrdp-sesman.log-unstable**
```
[20221116-14:03:30] [INFO ] starting xrdp-sesman with pid 886
[20221116-14:03:43] [INFO ] Socket 8: AF_INET6 connection received from ::1 port 36820
[20221116-14:03:45] [ERROR] pam_authenticate failed: Authentication failure
[20221116-14:03:45] [INFO ] Username or password error for user: webadmin
[20221116-14:03:45] [ERROR] sesman_data_in: scp_process_msg failed
[20221116-14:03:45] [ERROR] sesman_main_loop: trans_check_wait_objs failed, removing trans
[20221116-14:03:50] [INFO ] Socket 8: AF_INET6 connection received from ::1 port 53296
[20221116-14:03:50] [INFO ] Terminal Server Users group is disabled, allowing authentication
[20221116-14:03:50] [INFO ] ++ created session (access granted): username webadmin, ip ::ffff:192.168.1.100:54044 - socket: 12
[20221116-14:03:50] [INFO ] starting Xorg session...
[20221116-14:03:50] [INFO ] Starting session: session_pid 1649, display :10.0, width 1680, height 1050, bpp 24, client ip ::ffff:192.168.1.100:54044 - socket: 12, user name webadmin
[20221116-14:03:50] [INFO ] [session start] (display 10): calling auth_start_session from pid 1649
[20221116-14:03:50] [ERROR] sesman_data_in: scp_process_msg failed
[20221116-14:03:50] [ERROR] sesman_main_loop: trans_check_wait_objs failed, removing trans
[20221116-14:03:50] [INFO ] Starting X server on display 10: /usr/lib/xorg/Xorg :10 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log  
[20221116-14:03:50] [INFO ] Found X server running at /tmp/.X11-unix/X10
[20221116-14:03:50] [INFO ] Session started successfully for user webadmin on display 10
[20221116-14:03:50] [INFO ] Found X server running at /tmp/.X11-unix/X10
[20221116-14:03:50] [INFO ] Starting the xrdp channel server for display 10
[20221116-14:03:50] [INFO ] Session in progress on display 10, waiting until the window manager (pid 1663) exits to end the session
[20221116-14:03:50] [INFO ] Found X server running at /tmp/.X11-unix/X10
[20221116-14:03:50] [INFO ] Starting the default window manager on display 10: /etc/xrdp/startwm.sh
[20221116-14:05:00] [INFO ] shutting down sesman 1
[20221116-14:05:00] [INFO ] shutting down sesman 1
[20221116-14:05:24] [INFO ] shutting down sesman 1
[20221116-14:05:24] [WARN ] receiving SIGHUP 1
[20221116-14:05:24] [WARN ] Window manager (pid 1663, display 10) exited with non-zero exit code 255 and signal 1. This could indicate a window manager config problem
[20221116-14:05:24] [INFO ] Calling auth_stop_session and auth_end from pid 1649
[20221116-14:05:24] [INFO ] Terminating X server (pid 1664) on display 10
[20221116-14:05:24] [INFO ] Terminating the xrdp channel server (pid 1733) on display 10
[20221116-14:05:24] [INFO ] X server on display 10 (pid 1664) returned exit code 0 and signal number 0
[20221116-14:05:24] [INFO ] xrdp channel server for display 10 (pid 1733) exit code 0 and signal number 0
[20221116-14:05:24] [INFO ] cleanup_sockets:
```

**Comment by metalefty** [comment]
Can you try to increase the log level to DEBUG and try the same thing again?

BTW, I'm using xrdp with LE certificate with no problem.

**Comment by metalefty** [comment]
One more thing, can you configure xrdp to use the snakeoil certificate and see what happens? (with DEBUG log level)
As far as I see the log attached, your former setup didn't using the snakeoil certificate. It wasn't using TLS at all.

**Comment by iblackwell** [comment]
Will do, re. both your messages.  Fortunately, I snapshot'd the disk just before installing the LE cert, so I can roll back to the exact same "before" state.

Perhaps related to what you mention about the certs and for reasons unknown to me, it appears to me from the logs that an IPv4 connection is established using /etc/xrdp/cert.pem and then an IPv6 session is established, perhaps supplanting the IPv4?  Regardless, I found it odd that installing a cert for apache2 would affect a cert for xrdp.

Anyways, I'm on the info you requested.

**Comment by metalefty** [comment]
Your log implies that xrdp failed to load certificate's private key so xrdp fell into non-TLS mode (implicitly) as fallback.
That's why I think your before state were not using TLS. Your xrdp.ini is configured as `security_layer=negotiate`, right?

So, the difference between the before state and the after state is "not using TLS" vs "using TLS (with LE certificate)".
Not "snakeoil certificate" vs "LE certificate".

**xrdp.log-stable**
```
[20221116-12:38:14] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20221116-12:38:14] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20221116-12:38:14] [ERROR] Cannot read private key file /etc/xrdp/key.pem: Permission denied
```

**Comment by matt335672** [comment]
These lines aren't good:-

```
[20221116-14:05:00] [INFO ] shutting down sesman 1
[20221116-14:05:00] [INFO ] shutting down sesman 1
[20221116-14:05:24] [INFO ] shutting down sesman 1
[20221116-14:05:24] [WARN ] receiving SIGHUP 1
```

In the version of xrdp you've got, sesman doesn't handle SIGHUP properly, and is likely to fail (#1729). This was fixed for v0.9.19 by #2168.

If sesman fails, you lose the ability to reconnect to existing sessions even if sesman is restarted. We're working on a fix for that, but it needs a lot of changes.

Is there anything in your stack you're aware of which could be sending signals to sesman?

**Comment by iblackwell** [comment]
@metalefty, I understand what you are saying.  As you noted, in the 'stable' (before LE install) environment, xrdp can't read the key file.  In the 'unstable' (after LE install) environment, I had added xrdp to the ssl-cert group, which then removed that error.   To ensure that this was not the problem, I removed xrdp from the ssl-cert group and re-ran the test.  It still crashed.

Following up on your request for logs in DEBUG mode, attached are both the xrdp.log and xrdp-sesman.log files from both the stable and the unstable environment.

[xrdp.log.stable-221117.txt](https://github.com/neutrinolabs/xrdp/files/10034988/xrdp.log.stable-221117.txt)
[xrdp.log-unstable-221117.txt](https://github.com/neutrinolabs/xrdp/files/10034989/xrdp.log-unstable-221117.txt)
[xrdp-sesman.log.stable-221117.txt](https://github.com/neutrinolabs/xrdp/files/10034990/xrdp-sesman.log.stable-221117.txt)
[xrdp-sesman.log-unstable-221117.txt](https://github.com/neutrinolabs/xrdp/files/10034992/xrdp-sesman.log-unstable-221117.txt)

**Comment by iblackwell** [comment]
@matt335672, I can't think of anything that would interfere with xrdp.  That said, I did keep a change log as I put this server together.   A copy is attached (some minor redaction for security).

[WebHost-ChangeLog-221117.txt](https://github.com/neutrinolabs/xrdp/files/10035091/WebHost-ChangeLog-221117.txt)

**Comment by iblackwell** [comment]
A bit more info... Here is the part of the DEBUG log from xfreerdp where the session crashes:

[13:10:21:969] [17872:17873] [INFO][com.freerdp.core] - ERRINFO_LOGOFF_BY_USER (0x0000000C):The disconnection was initiated by the user logging off their session on the server.
[13:10:21:970] [17872:17873] [ERROR][com.freerdp.core] - rdp_set_error_info:freerdp_set_last_error_ex ERRINFO_LOGOFF_BY_USER [0x0001000C]
[13:10:21:970] [17872:17873] [DEBUG][com.freerdp.core.rdp] - DisconnectProviderUltimatum: reason: 3
[13:10:21:970] [17872:17873] [DEBUG][com.freerdp.core.rdp] - transport_check_fds() - -1
[13:10:21:970] [17872:17873] [DEBUG][com.freerdp.core] - rdp_check_fds() - -1

**Comment by matt335672** [comment]
Again, it looks like both processes are receiving unwanted signals:-

```
[20221117-10:29:46] [DEBUG] Received termination signal, stopping the client message processor thread
[20221117-10:29:46] [INFO ] Received termination signal, stopping the server accept new connections thread
[20221117-10:29:46] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:192.168.1.131 port 14403)
[20221117-10:29:46] [DEBUG] Closed socket 11 (AF_INET6 :: port 14403)
[20221117-10:29:46] [DEBUG] xrdp_mm_module_cleanup
[20221117-10:29:46] [DEBUG] Closed socket 19 (AF_UNIX)
[20221117-10:29:46] [DEBUG] Closed socket 20 (AF_UNIX)
```

and:-

```
[20221117-10:29:46] [INFO ] shutting down sesman 1
[20221117-10:29:46] [INFO ] shutting down sesman 1
[20221117-10:29:56] [INFO ] shutting down sesman 1
[20221117-10:29:56] [WARN ] receiving SIGHUP 1
```

Have you got journal logs for that time? Can you see what else might be running then?

Failing that, we can trace the signaller by using the audit mechanism:-

https://www.thegeekdiary.com/how-to-find-which-process-is-killing-mysqld-with-sigkill-or-sigterm-on-linux/

Have you configured certbot hooks to try to restart xrdp / sesman? They could be causing this.

**Comment by iblackwell** [comment]
I've attached a new xrdp.log, xrdp-sesman.log and journal.log, all trimmed of debris before and after the xrdp session.   While the SIGHUP happens at 11:05:41, my session is long gone by that point.  My session end at about 11:05:13 where, in xrdp-sesman.log it says, "shutting down sesman 1".

I will also take a look at the link you provided and see what that provides.
Thank-you.

[xrdp.log](https://github.com/neutrinolabs/xrdp/files/10044306/xrdp.log)
[xrdp-sesman.log](https://github.com/neutrinolabs/xrdp/files/10044309/xrdp-sesman.log)
[journal.log](https://github.com/neutrinolabs/xrdp/files/10044310/journal.log)

**Comment by iblackwell** [comment]
Another bit of information... Because this issue started immediately after installing the snap-based _certbot_ package, I decided to try un-installing certbot.  xrdp was immediately stable again; my sessions no longer crap-out.   I then installed the stand-alone certbot package (python3-based), and xrdp is still running solid.  So it appears, at least in my instance, there seems to be something afoul with the _ménage à trois_ of xrdp, certbot and snap.

**Comment by matt335672** [comment]
Does the snap-based certbot provide any logs of what it's doing?

**Comment by iblackwell** [comment]
There are no log entries for certbot at or around the time the xrdp session crashed.  However, you'll notice in the _journal.log_ file there are a number of snapd log entries that don't look good (e.g,. segfaults), including one entry right about the time the session drop off my screen that references certbot specifically  (noted in the most recent journal.log file I uploaded yesterday, and immediately below) @ 11:05;13:

_Nov 18 11:05:13 webhost snapd[782]: storehelpers.go:748: cannot refresh: snap has no updates available: "bare", "certbot", "core", "core20", "firefox", "gnome-3-38-2004", "gtk-common-themes", "snapd", "snapd-desktop-integration"_

So I'm leaning heavily in favour of pointing to snapd as the culprit (especially considering replacing the snap version of certbot with the non-snap version cleared things up).  I'm going to wander over to SnapCraft with the journal.log to see if they have any insight.

I'll report back here for completeness before closing this issue.

**Comment by matt335672** [comment]
@iblackwell - did you find out any more about this?

Thanks.

**Comment by iblackwell** [comment]
@matt335672  After my last post here I hit the SnapCraft forum with the same problem description and data.  I received one inquiry about how xrdp was being started, to which I responded. That was on Nov. 24.   The thread has been dark since then.

I removed the snap for xrdp and then installed xrdp directly.   I have not had a problem since then.   Which certainly lends support to ny hyopthesis noted above suggesting this is more of a snap problem than a xrdp problem.   So this issue has been resolved as far as xrdp is concerned.   SnapCraft is apparently uninterested in this issue so my thread will atrophy there.

Anyways, thank-you to you and @metalefty for getting engaged in this.  If nothing else it helped narrow this issue down to a Snap issue.   Much appreciated.
